lib/Analysis/Membar.cpp

```diff
@@ -35,22 +35,37 @@ void MembarAnalysis::resolve(FunctionOpInterface funcOp,
         return;
       }
     }
-    if (block->isEntryBlock())
+    // Start the analysis from the entry blocks of any nested isolated from
+    // above regions.
+    if (block->getParentOp()->hasTrait<OpTrait::IsIsolatedFromAbove>() &&
+        block->isEntryBlock())
       blockList.emplace_back(block);
   });
 
+  // `ttg.warp_specialize` is the only operation with regions present in the IR.
+  // Its default region is not isolated but is trivially branched to and from
+  // the parent op. Transparently iterate into the region within the context of
+  // the surrounding basic block.
+  auto advance = [](Operation *op) {
```

Comment:
Similarly, I wonder if it could be more general through any interface instead of special handling?

Reply:
This one is tricky because the region is essentially being treated as part of the parent basic block. RegionBranchOpInterface doesn't quite work because basic blocks aren't supposed to branch. The other option is to rewrite this analysis slightly to work on RegionBranchOpInterface, which I could do if you prefer that.

Reply:
> This one is tricky because the region is essentially being treated as part of the parent basic block.

I get it now. Let's keep the current structure

Reply:
I just realized this won't work because the default region can have a CFG inside of it 🤦. It should be relatively easy to rewrite this to handle a mix of CFG and region branch op interface. Let me give it a quick try.

Reply:
Okay it was a relatively simple change, but it seems the transfer function here isn't monotonic and the fixed point iteration can sometimes get stuck. Going to try a tentative fix is another PR just to kick ci

Reply:
https://github.com/triton-lang/triton/pull/5973

Reply:
> I just realized this won't work because the default region can have a CFG inside of it 🤦. It should be relatively easy to rewrite this to handle a mix of CFG and region branch op interface. Let me give it a quick try.

The CFG would be lowered to branches though, so there shouldn't be region branches

Reply:
I mean a literal CFG inside the default region, so something like

```
^bb(N>0):
  ttg.warp_specialize()
  default{
    cf.br ...
  }
```

I forgot to add a test for this.